file_reader.go

```diff
@@ -184,6 +209,25 @@ func (f *FileReader) Read(b []byte) (int, error) {
 		return 0, io.ErrClosedPipe
 	}
 
+	offset := f.offset
+	n, err := f.readImpl(b)
+
+	// Decrypt data chunk if file from HDFS encrypted zone.
+	if f.enc != nil && n > 0 {
+		plaintext, err := aesCtrStep(offset, f.enc, b[:n])
+		if err != nil {
+			f.offset = offset
+			return 0, err
+		}
+		for i := 0; i < n; i++ {
+			b[i] = plaintext[i]
+		}
+	}
+
+	return n, err
```

Comment:
`n` is now the length of the ciphertext, not the plaintext. It's important to return the amount of plaintext we wrote into `b`. It's weird that this doesn't cause test failures - I would try to reproduce a pathological case in a test before fixing the bug.

Reply:
See my comment below about block ciphers and assuming that len(in) ==len(out). Even if that is true in all the cases we can test, we shouldn't assume it's the case.

Reply:
aes-ctr (the only mode supported) is a stream cipher, not block one.